framework/filters/PageCache.php

```diff
@@ -213,43 +237,71 @@ protected function restoreResponse($response, $data)
      */
     public function cacheResponse()
     {
-        $this->view->endCache();
-        $response = Yii::$app->getResponse();
+        array_pop($this->view->cacheStack);
+        $beforeResult = $this->beforeCacheResponse();
+        if ($beforeResult === false) {
+            $content = ob_get_clean();
+            if (empty($this->view->cacheStack) && !empty($this->dynamicPlaceholders)) {
+                $content = $this->updateDynamicContent($content, $this->dynamicPlaceholders);
+            }
+            echo $content;
+            return;
+        }
+
+        $response = Yii::$app->response;
         $data = [
-            'format' => $response->format,
-            'version' => $response->version,
-            'statusCode' => $response->statusCode,
-            'statusText' => $response->statusText,
+            'cacheVersion' => 1,
+            'cacheData' => is_array($beforeResult) ? $beforeResult : null,
+            'content' => ob_get_clean()
         ];
-        if (!empty($this->cacheHeaders)) {
-            $headers = $response->getHeaders()->toArray();
-            if (is_array($this->cacheHeaders)) {
-                $filtered = [];
-                foreach ($this->cacheHeaders as $name) {
-                    $name = strtolower($name);
-                    if (isset($headers[$name])) {
-                        $filtered[$name] = $headers[$name];
-                    }
-                }
-                $headers = $filtered;
-            }
-            $data['headers'] = $headers;
+        if ($data['content'] === false || $data['content'] === '') {
+            return;
         }
-        if (!empty($this->cacheCookies)) {
-            $cookies = $response->getCookies()->toArray();
-            if (is_array($this->cacheCookies)) {
-                $filtered = [];
-                foreach ($this->cacheCookies as $name) {
-                    if (isset($cookies[$name])) {
-                        $filtered[$name] = $cookies[$name];
+
+        $data['dynamicPlaceholders'] = $this->dynamicPlaceholders;
+        foreach (['format', 'version', 'statusCode', 'statusText'] as $name) {
+            $data[$name] = $response->{$name};
+        }
+        foreach (['headers', 'cookies'] as $name) {
+            $property = 'cache' . ucfirst($name);
+            if ($this->{$property} !== false) {
+                $all = $response->{$name}->toArray();
+                if (is_array($this->{$property})) {
+                    $filtered = [];
+                    foreach ($this->{$property} as $subName) {
+                        if ($name === 'headers') {
+                            $subName = strtolower($subName);
+                        }
+                        if (isset($all[$subName])) {
```

Comment:
Conditional nesting is too deep and it makes hard to understand what's going on here.
Could be split to a separate private method with self-documenting name.